functions/cache/src/async_cache.cpp

```diff
@@ -0,0 +1,364 @@
+      KeyRequest request;
+      request.ParseFromString(serialized);
+
+      for (KeyTuple tuple : request.tuples()) {
+        Key key = tuple.key();
+
+        if (!tuple.has_lattice_type()) {
+          log->error("Cache requires type to retrieve key.");
+          continue;
+        } else if ((key_type_map.find(key) != key_type_map.end()) &&
+                   (key_type_map[key] != tuple.lattice_type())) {
+          log->error("lattice type mismatch.");
+          continue;
+        }
+
+        update_cache(key, tuple.lattice_type(), tuple.payload(),
+                     local_lww_cache, local_set_cache, log);
+        client.put_async(key, tuple.payload(), tuple.lattice_type(),
+                         request.response_address());
+      }
+    }
+
+    // handle updates received from the KVS
+    if (pollitems[2].revents & ZMQ_POLLIN) {
+      string serialized = kZmqUtil->recv_string(&update_puller);
+      KeyRequest updates;
+      updates.ParseFromString(serialized);
+
+      for (const KeyTuple& tuple : updates.tuples()) {
+        Key key = tuple.key();
+
+        // if we are no longer caching this key, then we simply ignore updates
+        // for it because we received the update based on outdated information
+        if (key_type_map.find(key) == key_type_map.end()) {
+          continue;
+        }
+
+        if (key_type_map[key] != tuple.lattice_type()) {
+          // This is bad! This means that we have a certain lattice type stored
+          // locally for a key and that is incompatible with the lattice type
+          // stored in the KVS. This probably means that something was put here
+          // but hasn't been propagated yet, and something *different* was put
+          // globally. I think we should just drop our local copy for the time
+          // being, but open to other ideas.
+
+          log->error("lattice type mismatch during key update from KVS.");
+
+          switch (key_type_map[key]) {
+            case LatticeType::LWW: local_lww_cache.erase(key); break;
+            case LatticeType::SET: local_set_cache.erase(key); break;
+            default:
+              break;  // this can never happen
+          }
+
+          key_type_map[key] = tuple.lattice_type();
+        }
+
+        update_cache(key, tuple.lattice_type(), tuple.payload(),
+                     local_lww_cache, local_set_cache, log);
+      }
+    }
+
+    vector<pair<KeyResponse, Address>> responses =
+        client.receive_async(kZmqUtil);
+    for (const auto& response : responses) {
+      Key key = response.first.tuples(0).key();
+      // TODO: assert that key_type_map[key] and
+      // response.first.tuples(0).lattice_type() are the same first, check if
+      // the request failed
+      if (response.first.response_id() == "NULL_ERROR") {
+        if (response.first.type() == RequestType::GET) {
+          client.get_async(key);
+        } else {
+          if (response.second != "") {
```

Comment:
all normal PUT should have `response_address` field, but the metadata update won't have this (in the current design). I'm not retrying because after `T` sec we will have another metadata update anyway

Reply:
Ah, okay, makes sense. Just add a comment making that clear.